data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java

```diff
@@ -74,111 +74,151 @@ private TableMigrationUtil() {
   public static List<DataFile> listPartition(Map<String, String> partition, String uri, String format,
                                              PartitionSpec spec, Configuration conf, MetricsConfig metricsConfig,
                                              NameMapping mapping) {
+    int parallelism = conf.getInt(PARQUET_READ_PARALLELISM, 1);
     if (format.contains("avro")) {
-      return listAvroPartition(partition, uri, spec, conf);
+      return listAvroPartition(partition, uri, spec, conf, parallelism);
     } else if (format.contains("parquet")) {
-      return listParquetPartition(partition, uri, spec, conf, metricsConfig, mapping);
+      return listParquetPartition(partition, uri, spec, conf, metricsConfig, mapping, parallelism);
     } else if (format.contains("orc")) {
-      return listOrcPartition(partition, uri, spec, conf, metricsConfig, mapping);
+      return listOrcPartition(partition, uri, spec, conf, metricsConfig, mapping, parallelism);
     } else {
       throw new UnsupportedOperationException("Unknown partition format: " + format);
     }
   }
 
   private static List<DataFile> listAvroPartition(Map<String, String> partitionPath, String partitionUri,
-                                                  PartitionSpec spec, Configuration conf) {
+                                                  PartitionSpec spec, Configuration conf, int parallelism) {
     try {
       Path partition = new Path(partitionUri);
       FileSystem fs = partition.getFileSystem(conf);
-      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
-          .filter(FileStatus::isFile)
-          .map(stat -> {
-            InputFile file = HadoopInputFile.fromLocation(stat.getPath().toString(), conf);
-            long rowCount = Avro.rowCount(file);
-            Metrics metrics = new Metrics(rowCount, null, null, null, null);
-            String partitionKey = spec.fields().stream()
+      List<FileStatus> fileStatus = Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
+              .filter(FileStatus::isFile)
+              .collect(Collectors.toList());
+      DataFile[] datafiles = new DataFile[fileStatus.size()];
+      Tasks.Builder<Integer> task = Tasks.range(fileStatus.size())
```

Comment:
I think we could just use Tasks.foreach(fileStatuses) here, I don't think the index is important

Reply:
> I think we could just use Tasks.foreach(fileStatuses) here, I don't think the index is important

Thank for your review.  If  `Tasks.foreach(fileStatuses)` was used instead , I have no idea how to collect the result safely.  appreciate for your thoughts. Many thanks.

Reply:
Couldn't we just accumulate the results in a concurrent structure?

Reply:
> Couldn't we just accumulate the results in a concurrent structure?

The internal of concurrent structure still has lock.  In my opinion, if we could do it in a lockfree method, why should we use concurrent structure

Reply:
And The most import part,  using  `index` , we can keep the order of DataFile in  `datafiles`   same with  the FileStatus in fileStatus.

Reply:
I'm not sure that we really have a performance contention issue here or that the order of the file statuses really is something we need to preserve, but it's fine if you want to keep it that way.